service-base/src/main/java/org/eclipse/hono/service/auth/device/TenantDetailsDeviceCredentials.java

```diff
@@ -0,0 +1,73 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.service.auth.device;
+
+import java.util.Objects;
+
+import org.eclipse.hono.util.TenantObject;
+import org.eclipse.hono.util.TenantObjectContainer;
+
+/**
+ * Credentials, provided by a device for authentication, along with detailed tenant information.
+ */
+public class TenantDetailsDeviceCredentials extends AbstractDeviceCredentials implements TenantObjectContainer {
```

Comment:
why don't we simply add a TenantObject typed field and a corresponding getter to *AbstractDeviceCredentials*?

Reply:
That `tenantObject` field would be `null` in all the places that an `AbstractDeviceCredentials` object is currently used. For the one use case introduced in this PR where the `tenantObject` is not `null`, I'd rather use a separate class where `tenantObject` is always set.

Reply:
FMPOV this will be the standard use case for all adapters, won't it? We create the (adapter specific) DeviceCredentials in order to authenticate the device and then put its values to the (adapter specific) context. I do not see why we wouldn't *always* include the TenantObject in the credentials ...

Reply:
That's not how I had envisioned it at this point. In the existing Hono code, the DeviceCredentials objects are created for example in `AuthProvider` classes (e.g. `CredentialsApiAuthProvider`) as part of the authentication handling. These `AuthProvider` classes could of course be changed to use the TenantClientFactory there to fetch and add the TenantObject to the DeviceCredentials there. But the DeviceCredentials object created there isn't returned by the `AuthProvider` methods, namely the `authenticate` methods. So these `authenticate` methods would have to be changed there (or the returned `DeviceUser` would have to get a TenantObject as well or something like that).

The whole goal of this PR is to fetch the TenantObject early on, i.e. _before_ the authentication check, so that e.g. a connection limit check can be done before the (expensive) authentication check. But when using an approach with the tenant-enhanced DeviceCredentials in the `AuthProvider` classes, the TenantObject would be fetched as part of the authentication check, not before.
As a way around this, I could imagine using some kind of "preAuthenticationCheckHandler" parameter, used inside the "AuthProvider.authenticate" methods _after_ having fetched the credentials with tenantObject and _before_ calling "getCredentialsForDevice(deviceCredentials)" and "validateCredentials(deviceCredentials". Inside this "preAuthenticationCheckHandler" (given to the AuthProvider from the protocol adapter class) e.g. the connection limit check could be done.

But this would require further changes to the `AuthProvider` classes.
The advantage would of course be to not have the tenantId/authId extraction logic duplicated as it is now, performing it once for getting the TenantObject and then for doing the authentication. I had thought about reusing the UsernamePasswordAuthProvider/X509AuthProvider classes for that, but found this not to be a good solution. Hence my approach to with a more lightweight, pluggable tenantId/authId extraction logic (without certificate check for example) to be used alongside/before the authentication logic.

But changing `AuthProvider` instead and adding such a "preAuthenticationCheckHandler" mechanism for the "authenticate" method, now also sounds like a good solution, probably even a better one. Before doing that I would just want to see agreement, that we want to change the authentication interfaces in such a way. @sophokles73 WDYT?


(As a side note: in the AMQP adapter there is already a "preAuthenticationHandler" mechanism, there it's in the AmqpAdapterSaslAuthenticatorFactory.)

Reply:
Having looked deeper into that, I think for the MQTT adapter we would need to use different, new AuthHandler/AuthProvider methods that don't rely on the credentials being passed around between "parseCredentials()" and "authenticate()" in a JsonObject as intermediate format, with no possibility to add further non-String parameters (like e.g. the ExecutionContext). Otherwise the OpenTracing root span of the current MQTT CONNECT handling can't be made available to the "preAuthenticationCheckHandler".

Reply:
I've created #2140 to prepare the Hono `AuthHandler` for this, using the `ExecutionContext` in the `AuthProvider` `authenticate` method, so that the `ExecutionContext` can be given to the "preAuthenticationCheckHandler".

I now think we can use this in the MQTT and probably also AMQP adapter to get rid of the duplicate "get tenant/authId from auth credentials" logic there. For the HTTP adapter that would not be so straightforward, as that would require overwriting and copying `io.vertx.ext.web.handler.impl.AuthHandlerImpl.handle` to add the "preAuthenticationHandler" hook.

That means I would now go ahead with the `AuthProvider` approach and close this PR (applying parts of it later).